Frontend migration of portfolio-grade-select-step

Pasindufdo98 opened the pull request:
Frontend Migration: portfolio-grade-select-step

# Description

This component is used in the second step of the portfolio creation flow to allow users to select a grade (P, C, D, or HD) based on the assessment criteria they have demonstrated. In this PR, I have migrated the portfolio-grade-select-step component from AngularJS (CoffeeScript) to Angular 17 using TypeScript. The legacy .coffee and .tpl.html files have been removed, and the related references in doubtfire-angularjs.module.ts have been cleaned up.


## Type of change

- Migrated: portfolio-grade-select-step.coffee → portfolio-grade-select-step.component.ts, .html, .scss
- Deleted: portfolio-grade-select-step.coffee, portfolio-grade-select-step.tpl.html
- Updated: doubtfire-angularjs.module.ts (removed old directive reference)
- Updated: portfolio.tpl.html to mount the new Angular 17 component

# How Has This Been Tested?

Navigated to second step of the portfolio creation: Select Grade in the Portfolio Wizard and verified the following:

- Grade options (P, C, D, HD) are displayed with the correct styling and alignment.
- The checkbox labeled “I have read the Assessment Criteria for this unit” toggles correctly and enables grade selection only when checked.
- Clicking a grade triggers ProjectService.update() as expected, confirmed through network requests and browser console logs.
- On refresh, the selected grade is correctly persisted and reflected in the UI.

Browser DevTools (Console and Network tabs) were used to:

- Log project.submittedGrade before and after updates.
- Confirm that the correct payload is sent to the backend via ProjectService.update().
- Ensure no unexpected errors or warnings appear during component interaction.

## Screenshots

### Before migration

![before 1](https://github.com/user-attachments/assets/a693ce02-b7ab-4b72-a219-80db33d25976)

![before 2](https://github.com/user-attachments/assets/9ff576cc-2840-4ad2-a483-083943f582d5)

## After migration

![after 1](https://github.com/user-attachments/assets/ffe6d4bd-1777-46e6-805d-a00d9cb4bd41)

![after 2](https://github.com/user-attachments/assets/9d413c28-778a-45eb-bf19-1ecc353fb292)


## Testing Checklist:

- [x] Tested in latest Chrome
- [ ] Tested in latest Safari
- [x] Tested in latest Firefox

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have requested a review

Pasindufdo98 committed:
Frontend migration of portfolio-grade-select-step to Angular 17 component

lachlan-robinson reviewed (changes requested):
Hi @Pasindufdo98,

This migration appears to be **mostly complete** — the old component has been removed, and the new component has been successfully linked and downgraded. Great work so far!

There are a few **minor tweaks** I would recommend to polish the implementation:

##  Summary
- [ ] The component loads without errors.
- [ ] Functionality is consistent with the previous implementation.

## Component Logic and Code Quality
- [ ] Unused Inputs:
  - `unit` input is declared but not used.  
  - `project_data` input is declared but not used.
- [ ] Type Safety:  
  - `project` input is used but should be typed using the `Project` interface.
- [ ] Unusual Scope Injection:  
  - Injecting `$scope` and navigating based on it is not typical. Consider handling this via a service instead.
- [ ] Unused Imports:  
  - Clean up unused services/modules such as:  
    ```ts
    Inject, signal, ElementRef, AfterViewInit, angular, AlertService
    ```
- [ ] Duplicate/Redundant Imports:  
  - Imports can be consolidated and cleaned:
    ```ts
    import { Component, Input, Inject, signal } from '@angular/core';
    import { ElementRef, Injector, AfterViewInit } from '@angular/core';
    ```
- [ ] Variable Typing:  
  - `targetGrade` is declared as `null`. It should ideally have a proper type.
- [ ] Missing Interface Implementation:  
  - `PortfolioGradeSelectStepComponent` should implement `OnInit`.
- [ ] Error Handling:  
  - Consider using the `AlertService` to improve user feedback and error visibility.

## Template and Styling
- [ ] Icon Usage:  
  - The template currently uses Font Awesome. Replace with [Material UI icons](https://mui.com/material-ui/material-icons/) for consistency with Angular Material design.
- [ ] Styling Frameworks:  
  - Elements like `btn btn-primary` are still Bootstrap-based. Consider switching to Tailwind or SCSS-based styles.
- [ ] Layout Refinement:
  - Slight adjustment needed for spacing between the **acknowledgement** and the **checkbox**

![Screenshot 2025-05-05 104848](https://github.com/user-attachments/assets/7f23d5f5-49a8-4095-951a-99b0fdfffa3c)

Pasindufdo98 committed:
Addressed review comments in portfolio-grade-select-step component

Pasindufdo98 committed:
Implement OnInit in PortfolioGradeSelectStepComponent

Pasindufdo98 commented:
Hi @lachlan-robinson ,

Thanks for the feedback and I worked on the above suggestions and fixed followings.
- Cleaned up the component logic and removed unused or redundant initializations.
- Replaced targetGrade = null with an optional number (targetGrade?: number) to reflect that the variable will be assigned later during execution.
- Confirmed that the project and unit inputs are used in the template even though they're not directly referenced in the TypeScript logic.
- Added implements OnInit to make the component's lifecycle behavior more explicit and clear.
- Retained $scope.advanceActiveTab to maintain navigation behavior between steps (Original logic has been implemented in "portfolio" parent component which has to be migrated at the last)
- Removed AlertService that is unnecessary for this component.
- Adjusted spacing between the acknowledgement and the checkbox.

lachlan-robinson commented on src/app/projects/states/portfolio/directives/portfolio-grade-select-step/portfolio-grade-select-step.component.ts:
```diff
@@ -0,0 +1,65 @@
+  encapsulation: ViewEncapsulation.None,
+})
+export class PortfolioGradeSelectStepComponent implements OnInit {
+  @Input() project: any;
+  @Input() unit: any;
```
Add typing to the inputs?
```suggestion
  @Input() project: Project;
  @Input() unit: Unit;
```

lachlan-robinson reviewed (changes requested):
Nice work on implementing the changes! LGTM

chelaz1234 commented:
You've done a solid job with the code migration everything is working as expected, which is great to see. You've clearly specified ViewEncapsulation.None, showing good awareness of how styling should be applied across components. The component downgrading for AngularJS compatibility is also well-executed. On top of that, you've effectively implemented the changes Lachlan mentioned. Overall, great work!

atharv02-git reviewed (approved):
Hi @Pasindufdo98, just went through your code changes and the migration, here is my feedback:

1. Successfully unlinked and removed the older instances of the component from `doubtfire-angularjs.module.ts` and  `directives.coffee`.
2. Successfully imported the new component into the `doubtfire-angular.module.ts`.
3. Has successfully deleted old coffee files and new .ts instances of the components has been created with clear .component.ts, .component.html and .component.scss.
4. `portfolio-grade-select-step.component.html` clearly uses the latest html practices to align with `portfolio-grade-select-step.component.ts.
5. Successfully updated the property binding style to use the Angular form `portfolio.tpl.html`.
6. I was able to Pull your request and build your changes without any error.
7. I can also confirm that the changes @lachlan-robinson suggested for has been clearly implemented.

Migration Looks solid and well executed. LGTM. 
After Migration:
![image](https://github.com/user-attachments/assets/52d79b80-c661-46d9-9a80-bfa9b211a323)

aNebula commented:
LGTM.
Please open the upstream pull request with these changes against [doubtfire-lms/doubtfire-web 9.x branch](https://github.com/doubtfire-lms/doubtfire-web/tree/9.x)
@Pasindufdo98

Pasindufdo98 committed:
replaced bootsrap UI component with Angular material
